lib/isodoc/presentation_function/section.rb

```diff
@@ -15,11 +15,25 @@ def clause(docxml)
     end
 
     def unnumbered_clause?(elem)
-      !elem.ancestors("boilerplate, metanorma-extension").empty? ||
+      has_required_ancestor?(elem) ||
         @suppressheadingnumbers || elem["unnumbered"] ||
         elem.at("./ancestor::*[@unnumbered = 'true']")
     end
 
+    def has_required_ancestor?(elem)
+      @ancestor_cache ||= {}
+
+      return false unless elem
+      return @ancestor_cache[elem] if @ancestor_cache.key?(elem)
```

Comment:
This implementation is strange: making the cache key a Nokogiri element is hardly going to be efficient, although mercifully this is only clauses, so the cache will not get too big. (If it did, you can rely on clauses to have an id attribute.) has_required_ancestor? is a bad choice of name, this is specific to whether a clause counts as unnumbered (numbering is irrelevant in boilerplate and metanorma-extensions), so I will rename the method.